Posts/2021/09/my-crescendo-journey.md

```diff
@@ -0,0 +1,158 @@
+                    "Handler": "ParseProvider"
+                }
+            ],
+         }
+    ]
+}
+```
+
+The `ParseProvider` function is one of the functions that I had written to parse the output. I
+repeated this pattern to create a new cmdlet for each of the `vssadmin` subcommands.
+
+Notice that the first line of the JSON references a schema file. This file comes with the Crescendo
+module. I copied that to the folder where I was developing my module. I used Visual Studio Code (VS
+Code) to do all my development. With this schema file, VS Code provides IntelliSense for the JSON,
+making it easy to know what values are required and the type of information needed.
+
+Eventually, I added properties to the configuration for full help with descriptions and examples.
+And I defined parameter sets for the `vssadmin` commands that supported parameters.
+
+## Creating the new module
+
+Crescendo, itself, is a module. It contains cmdlets that help you create your configuration and then
+uses that configuration to create the module containing your cmdlets. Once I was happy with the
+configuration file, I used the `Export-CrescendoModule` cmdlet to create my module.
+
+```powershell
+Export-CrescendoModule -ConfigurationFile .\vssadmin.crescendo.config.json -ModuleName VssAdmin.psm1
+```
+
+Crescendo created two new files:
+
+- The module code file `VssAdmin.psm1`
+- The module manifest file `VssAdmin.psd1`
+
+These are the only two files that need to be installed. The `VssAdmin.psm1` file contains all the
+cmdlets that Crescendo generated from the configuration and the **Output Handler** functions I
+wrote to parse the output into objects.
+
+The end result was a well-structured, fully documented module.
+
+I still have one cmdlet left to create and I need to add administrative elevation since `vssadmin`
+requires it. But I am happy with the results I have so far.
+
+## Conclusion
+
+After reading all of this you might still be asking "how is this any easier than just writing the
+module myself?"
+
+That is a fair question. But here are the conclusions I came to as I went through this process.
+
+- The whole process, starting from nothing, researching both Crescendo and `vssadmin`, writing the
+  code, creating the configuration, and generating the module took me about 4 hours. I thought that
+  was pretty fast.
+- Crescendo lets you separate the logic code (your parsing functions) from the cmdlet definition and
+  parameter handling code. I found it easier to describe the cmdlets and their parameters in the
+  JSON file rather than having to write that code myself.
+- Crescendo handles things like **CommonParameters** and `SupportsShouldProcess` for you. You don't
+  have to write that support code in the cmdlets.
+- Separating the declarative code (the JSON configuration) from the logical code (your parsers)
+  makes is easier to add functionality to your module if the native command-line tool is updated.
+
```

Comment:
it can also create the comment based help with everything filled in. no need to go looking up that format either.

Reply:
Good point. Will add it.